src/utils/hooks/useGetResponsesAndElementState.tsx

```diff
@@ -0,0 +1,137 @@
+import { useState, useEffect } from 'react'
+import {
+  useGetElementsAndResponsesQuery,
+  GetElementsAndResponsesQuery,
+  ApplicationResponse,
+  TemplateSection,
+  TemplateElement,
+} from '../generated/graphql'
+import {
+  ResponsesByCode,
+  ResponsesFullByCode,
+  TemplateElementState,
+  EvaluatedElement,
+} from '../types'
+import evaluateExpression from '@openmsupply/expression-evaluator'
+
+const useGetResponsesAndElementState = (props: { serialNumber: string }) => {
+  const { serialNumber } = props
+  const [responsesByCode, setResponsesByCode] = useState({})
+  const [responsesFullByCode, setResponsesFullByCode] = useState({})
+  const [elementsExpressions, setElementsExpressions] = useState([{}])
+  const [elementsState, setElementsState] = useState({})
+  const [error, setError] = useState('')
+  const [loading, setLoading] = useState(true)
+  const { data, loading: apolloLoading, error: apolloError } = useGetElementsAndResponsesQuery({
+    variables: { serial: serialNumber },
+  })
+
+  useEffect(() => {
+    const error = checkForApplicationErrors(data)
+
+    if (error) {
+      setError(error)
+      setLoading(false)
+      return
+    }
+
+    if (apolloError) return
+
+    const applicationResponses = data?.applicationBySerial?.applicationResponses
+      .nodes as ApplicationResponse[]
+
+    const templateSections = data?.applicationBySerial?.template?.templateSections
+      .nodes as TemplateSection[]
+
+    const templateElements: TemplateElementState[] = []
+
+    templateSections.forEach((sectionNode) => {
+      const elementsInSection = sectionNode.templateElementsBySectionId?.nodes as TemplateElement[]
+      elementsInSection.forEach((element) => {
+        templateElements.push(element as TemplateElementState) // No idea why ...[spread] doesn't work here.
+      })
+    })
+
+    const currentResponses = {} as ResponsesByCode
+    const currentFullResponses = {} as ResponsesFullByCode
+
+    applicationResponses.forEach((response) => {
+      const code = response?.templateElement?.code
+      if (code) {
+        currentResponses[code] = response?.value?.text
+        currentFullResponses[code] = response?.value
+      }
+    })
+
+    setResponsesByCode(currentResponses)
+    setResponsesFullByCode(currentFullResponses)
+    setElementsExpressions(templateElements)
+    setLoading(false)
```

Comment:
I've found what was causing the problem to refresh the page! This `setLoading` should be moved to the next useEffort. I'll add one suggestion next.

Reply:
Great! Thanks for picking up on this. :)